### ISSUE TITLE
QTree issues

### ISSUE THREAD

**Issue by mjmare** [elimintz/justpy#80]
Continuing my exploration of Quasar components: QTree. 

Observations FYI (but might all be already known):

- selection, ticked, expanded will not work directly because of the two-way sync requirement. However it can be more or less worked around by using event handlers.
- content customisation won't work due to the use of slots with props.
- for some reason setting the default_expand_all property to False (or 'false') does not work
- self.run_method('expandAll ()', msg.websocket) does not work

My code:
```
class MoviesTreePage(BasePage):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        self.nodes = []
        self.expanded = []
        self.selected = None

        async def on_click0(comp, msg):
            # does not work yet
            await self.run_method('expandAll ()', msg.websocket)
            await self.update()

        async def on_click1(comp, msg):
            self.tree.expanded = []
            await self.update()

        async def on_click2(comp, msg):
            self.tree.expanded = ['B']
            await self.update()

        self.btn0 = jp.QBtn(a=self, text='Expand all', click=on_click0)
        self.btn1 = jp.QBtn(a=self, text='Close all', click=on_click1)
        self.btn2 = jp.QBtn(a=self, text='Open B', click=on_click2)


        async def on_expanded(comp, msg):
            logger.info('EXPANDED %s', msg.value)
            self.expanded = msg.value

        async def on_selected(comp, msg):
            logger.info('SELECTED %s', msg.value)
            self.selected = msg.value

        self.tree = jp.QTree(a=self,
                             # nodes=self.nodes,
                             node_key='label',
                             # expanded=self.expanded,
                             selected_color="primary",
                             default_expand_all=False,
                             )
        self.tree.on('update:expanded', on_expanded)
        self.tree.on('update:selected', on_selected)

        self.load_data()

    def load_data(self):
        self.nodes = [
            {'label': 'A',
             'icon': 'restaurant_menu',
             'children': [
                 {'label': 'Prompt attention'},
                 {'label': 'Professional waiter'}
             ]},
            {'label': 'B',
             'icon': 'photo',
             'children': [
                 {'label': 'Good table presentation', 'icon': 'restaurant_menu',},
                 {'label': 'Pleasing decor'}
             ]},
            {'label': 'C',
             'icon': 'room_service',
             'disabled': True,
             'children': []},

        ]

        self.tree.nodes = self.nodes
```

**Comment by elimintz** [comment]
I think the problem may be that you have an extra space in:
`await self.run_method('expandAll ()', msg.websocket)`
Should be:
`await self.run_method('expandAll()', msg.websocket)`

This program works for me:
```python
import justpy as jp

async def expand_tree(self, msg):
    return await self.tree.run_method('expandAll()', msg.websocket)

async def collapse_tree(self, msg):
    return await self.tree.run_method('collapseAll()', msg.websocket)


def tree_test():
    wp = jp.QuasarPage()
    d = jp.Div(classes="q-pa-md q-gutter-sm", a=wp)
    node_string = """
    [
        {
          label: 'Satisfied customers (with avatar)',
          avatar: 'https://cdn.quasar.dev/img/boy-avatar.png',
          children: [
            {
              label: 'Good food (with icon)',
              icon: 'restaurant_menu',
              children: [
                { label: 'Quality ingredients', icon: 'favorite' },
                { label: 'Good recipe' }
              ]
            },
            {
              label: 'Good service (disabled node with icon)',
              icon: 'room_service',
              disabled: true,
              children: [
                { label: 'Prompt attention' },
                { label: 'Professional waiter' }
              ]
            },
            {
              label: 'Pleasant surroundings (with icon)',
              icon: 'photo',
              children: [
                {
                  label: 'Happy atmosphere (with image)',
                  img: 'https://cdn.quasar.dev/img/logo_calendar_128px.png'
                },
                { label: 'Good table presentation' },
                { label: 'Pleasing decor' }
              ]
            }
          ]
        }
      ]
    """

    b1 = jp.QBtn(label='Expand', a=d, click=expand_tree)
    b2 = jp.QBtn(label='Collapse', a=d, click=collapse_tree)

    tree = jp.QTree(a=d, node_key='label', nodes=node_string, tick_strategy="leaf", no_connectors=False, default_expand_all=False)
    d1 = jp.Div(text='', a=d)
    def my_updated(self, msg):
        print('in my updated')
        d1.text = str(msg.value)
    tree.on('update:ticked', my_updated)
    b1.tree = tree
    b2.tree = tree
    return wp

jp.justpy(tree_test)
```

Please let me know if this does not solve the problem.

As for default_expand_all, i need to look into it. Thanks for catching this.

**Comment by mjmare** [comment]
No it was not the extraneous space; javascript accepts all kinds of weirdness ;-)
But your example pointed me in the right direction: I should have written return 
```
await self.tree.run_method('expandAll()', msg.websocket)
```
instead of:
```
return await self.run_method('expandAll()', msg.websocket)
```
Thx!

**Comment by mjmare** [comment]
BTW would it be possible to support methods like expandAll by default? So that a user would have to do the run_method thing?

**Comment by elimintz** [comment]
Thank you, I didn't know that about JavaScript....
Now that I look at it again, I think the problem is the page update that you also had. That reset the page to its previous state.

Yes, I will add the methods. There are so many of them, so it will take some time.

**Comment by mjmare** [comment]
> Thank you, I didn't know that about JavaScript....

It is usually the linter that fussy. I usually put his at the end of Quasar's eslintrc.js:
```
    // MJM
    'indent': 0,
    'no-multiple-empty-lines': ["error", { "max": 2, "maxEOF": 1 }],
    'comma-dangle': 0,
    'semi': 0,
    'prefer-const': 0
```
Makes javascript much more palatable. 
> Now that I look at it again, I think the problem is the page update that you also had. That reset the page to its previous state.

No, I have the component update still in, and it works. The call is superfluous but does not harm either. 

> 
> Yes, I will add the methods. There are so many of them, so it will take some time.

Need help?

I suppose only methods that do not return anything can be supported at this time, right?

**Comment by elimintz** [comment]
Yes, I would appreciate help with this. 
The return value of the methods is ignored at this time, though they will be run.

**Comment by elimintz** [comment]
The default_expand_all bug is fixed in the new version 0.1.2

**Comment by mjmare** [comment]
Thx
